src/Symfony/Component/JsonPath/JsonCrawler.php

```diff
@@ -562,16 +824,260 @@ private function evaluateRecursive(mixed $value): array
     private function compare(mixed $left, mixed $right, string $operator): bool
     {
         return match ($operator) {
-            '==' => $left === $right,
-            '!=' => $left !== $right,
-            '>' => $left > $right,
-            '>=' => $left >= $right,
-            '<' => $left < $right,
-            '<=' => $left <= $right,
+            '==' => $this->compareEquality($left, $right),
+            '!=' => !$this->compareEquality($left, $right),
+            '>', '>=', '<', '<=' => $this->compareOrdering($left, $right, $operator),
             default => false,
         };
     }
 
+    private function compareEquality(mixed $left, mixed $right): bool
+    {
+        $leftIsNothing = $left === self::nothing();
+        $rightIsNothing = $right === self::nothing();
+
+        if (
+            $leftIsNothing && $rightIsNothing
+            || ($leftIsNothing && 0 === $right || 0 === $left && $rightIsNothing)
+        ) {
+            return true;
+        }
+
+        if ($leftIsNothing || $rightIsNothing) {
+            return false;
+        }
+
+        if ((\is_int($left) || \is_float($left)) && (\is_int($right) || \is_float($right))) {
+            return $left == $right;
+        }
+
+        if (\is_string($left) && \is_string($right) || \is_bool($left) && \is_bool($right)) {
+            return $left === $right;
+        }
+
+        if (null === $left && null === $right) {
+            return true;
+        }
+
+        // arrays must have equal length and equal corresponding elements
+        if (\is_array($left) && \is_array($right)) {
+            return $this->compareArraysDeep($left, $right);
+        }
+
+        // objects must have identical names and equal corresponding values
+        if ($left instanceof \stdClass && $right instanceof \stdClass) {
+            return $this->compareObjectsDeep($left, $right);
+        }
+
+        // null (missing property) equals 0 when compared to function results
+        if (null === $left && 0 === $right || 0 === $left && null === $right) {
+            return true;
+        }
+
+        // different types are not equal
+        return false;
+    }
+
+    private function compareArraysDeep(array $left, array $right): bool
+    {
+        $leftIsList = array_is_list($left);
+        $rightIsList = array_is_list($right);
+        $leftCount = \count($left);
+
+        if ($leftIsList !== $rightIsList || $leftCount !== \count($right)) {
+            return false;
+        }
+
+        if ($leftIsList) {
+            for ($i = 0; $i < $leftCount; ++$i) {
+                if (!$this->compareEquality($left[$i], $right[$i])) {
+                    return false;
+                }
+            }
+        } else {
+            foreach ($left as $key => $value) {
```

Comment:
this foreach means the order of the keys don't matter - is this covered by the test suite?

Reply:
Actually we can keep only this loop, the comparison is order independent (see https://datatracker.ietf.org/doc/rfc9535/ section 2.3.5.2.2)